### PR TITLE
fix: amplitude fix for user operations

### PR DESCRIPTION
**Pull request by utsabc**
## What are the changes introduced in this PR?

The following changes address the inconsistencies observed for the `userOperations` 

## What is the related Linear task?

Resolves INT-1641

## Please explain the objectives of your changes below

In amplitude the `userOperations` were only subjected to `identify` flow but this is applicable for `page`, `track` and `screen` as these events can also contain `user_properties` in the payload. Along with that the following inconsistency was observed in the flow which is also fixed with post process function

Screenshot 2024-03-04 at 10 36 02 AM

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

The changes are introduced over configuration flag which prevents introducing this as a breaking change

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

## Summary by CodeRabbit

- **Refactor**
	- Renamed function for clarity in handling user properties.
- **New Features**
	- Introduced post-processing for user properties to enhance data handling.
	- Updated data response logic to support new user properties handling based on configuration.

### CODE REVIEW

**Comment by utsabc** [comment]
@coderabbitai review

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3153/test-report.html

File: src/v0/destinations/am/transform.js [hunk 1]
```diff
@@ -335,14 +335,65 @@ const getDefaultResponseData = (message, rawPayload, evType, groupInfo) => {
...
+  const { user_properties } = rawPayload;
+  const userPropertiesKeys = Object.keys(user_properties).filter(
+    (key) => !operationList.includes(key),
+  );
+  const duplicatekeys = new Set();
+  // eslint-disable-next-line no-restricted-syntax, guard-for-in
+  for (const key of userPropertiesKeys) {
+    // check if any of the keys are present in the user_properties $setOnce, $add, $unset, $append, $prepend, $preInsert, $postInsert, $remove keys as well as root level
+
+    if (
```
Comment (krishna2020) [inline src/v0/destinations/am/transform.js:360]: how can this scenario happen ?

File: src/v0/destinations/am/transform.js [hunk 2]
```diff
@@ -335,14 +335,65 @@ const getDefaultResponseData = (message, rawPayload, evType, groupInfo) => {
...
+    ) {
+      duplicatekeys.add(key);
+    }
+  }
+  // eslint-disable-next-line no-restricted-syntax, guard-for-in
+  for (const key of duplicatekeys) {
+    delete user_properties[key];
+  }
+
+  const setProps = {};
```
Comment (krishna2020) [inline src/v0/destinations/am/transform.js:373]: Add a comment
// Moving root level properties that are doesn't belong to any operation under $set

File: src/v0/destinations/am/transform.js [hunk 3]
```diff
@@ -357,8 +408,15 @@ const getResponseData = (evType, destination, rawPayload, message, groupInfo) =>
     case EventType.ALIAS:
       break;
     default:
+      if (destination.Config.enableEnhncedUserOpertaions) {
```
Comment (krishna2020) [inline src/v0/destinations/am/transform.js:411]:
```suggestion
      if (destination.Config.enableEnhancedUserOperations) {
```

File: src/v0/destinations/am/transform.js [hunk 4]
```diff
@@ -335,14 +335,65 @@ const getDefaultResponseData = (message, rawPayload, evType, groupInfo) => {
   const groups = groupInfo && cloneDeep(groupInfo);
   return { groups, rawPayload };
 };
+
```
Comment (krishna2020) [inline src/v0/destinations/am/transform.js:338]: move this new function to am/utils.js & have few basic unit tests

Comment (utsabc) [reply src/v0/destinations/am/transform.js:338]: added

File: src/v0/destinations/am/transform.js (hunk 3 above)
Comment (utsabc) [reply src/v0/destinations/am/transform.js]: fixed

File: src/v0/destinations/am/transform.js (hunk 2 above)
Comment (utsabc) [reply src/v0/destinations/am/transform.js]: added

**Review by ItsSudip — COMMENTED** [review]
Can you add some unit test case for component as well?

File: src/v0/destinations/am/transform.js (hunk 1 above)
Comment (krishna2020) [reply src/v0/destinations/am/transform.js]: Ideally this should happen in the respective function blocks then when we release this functionality to all customers

**Review by krishna2020 — APPROVED** [review]

**Comment by krishna2020** [comment]
@utsabc should have squash merged it as all commits belong to one feature

**Pull request merged**